src/repr/src/adt/array.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A multi-dimensional array data type.
+
+use std::convert::TryInto;
+use std::error::Error;
+use std::fmt;
+use std::mem;
+
+use serde::{Deserialize, Serialize};
+
+use crate::row::DatumList;
+
+/// The maximum number of dimensions permitted in an array.
+pub const MAX_ARRAY_DIMENSIONS: u8 = 6;
+
+/// A variable-length multi-dimensional array.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Array<'a> {
+    /// The dimensions of the array.
+    pub(crate) dims: ArrayDimensions<'a>,
+    /// The elements in the array.
+    pub(crate) elements: DatumList<'a>,
+}
+
+impl<'a> Array<'a> {
+    /// Returns the dimensions of the array.
+    pub fn dims(&self) -> ArrayDimensions<'a> {
+        self.dims
+    }
+
+    /// Returns the elements of the array.
+    pub fn elements(&self) -> DatumList<'a> {
+        self.elements
+    }
+}
+
+/// The dimensions of an [`Array`].
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct ArrayDimensions<'a> {
+    pub(crate) data: &'a [u8],
+}
+
+impl ArrayDimensions<'_> {
+    /// Returns the number of dimensions in the array as a [`u8`].
+    pub fn ndims(&self) -> u8 {
+        let ndims = self.data.len() / (mem::size_of::<usize>() * 2);
+        ndims.try_into().expect("ndims is known to fit in a u8")
+    }
+
+    /// Returns the number of the dimensions in the array as a [`usize`].
+    pub fn len(&self) -> usize {
+        self.ndims().into()
+    }
+
+    /// Reports whether the number of dimensions in the array is zero.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl fmt::Debug for ArrayDimensions<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_list().entries(self.into_iter()).finish()
+    }
+}
+
+impl<'a> IntoIterator for ArrayDimensions<'a> {
+    type Item = ArrayDimension;
+    type IntoIter = ArrayDimensionsIter<'a>;
+
+    fn into_iter(self) -> ArrayDimensionsIter<'a> {
+        ArrayDimensionsIter { data: self.data }
+    }
+}
+
+/// An iterator over the dimensions in an [`ArrayDimensions`].
+#[derive(Debug)]
+pub struct ArrayDimensionsIter<'a> {
+    data: &'a [u8],
+}
+
+impl Iterator for ArrayDimensionsIter<'_> {
+    type Item = ArrayDimension;
+
+    fn next(&mut self) -> Option<ArrayDimension> {
+        if self.data.is_empty() {
+            None
+        } else {
+            let lower_bound = unsafe { (self.data.as_ptr() as *const usize).read_unaligned() };
```

Comment:
https://doc.rust-lang.org/std/primitive.u32.html#method.from_ne_bytes should let you avoid the `unsafe`.

Reply:
Oh `ne` means "native endian" and not "network endian"? -_- Thanks, will adjust.

Reply:
yeah it's confusing if you know about things like `htonl` from C, but network/big endian is just `be` in these APIs. -_- indeed.

Reply:
Fixed.